### PR TITLE
Feature: document comments

### PR DESCRIPTION
**Pull request by tim-vogel**
## Proposed change

**new Feature list:**
- [x] - Tab for displaying comments in the document detail view.
- [x] - List all comments for a document in descending order (newest comments first)
- [x] - add new comments
- [x] - delete comments
- [ ] - ~~full text search over comments~~
- [x] -  integration of comments into the standard import and export function of paperless
- [x] - activate / deactivate the comment function (only the view is hidden or shown)
- [x] - new environment variable ```PAPERLESS_COMMENTS_ENABLED``` -> [true: enabled | false: disabled]
- [x] - new service to retrieve backend environment variables. 

By default the comment function is disabled. To enable it, the environment variable ```PAPERLESS_COMMENTS_ENABLED``` must be set to ```true```

A possible full text search for the comments would have to be added by an experienced programmer.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [ ] If applicable, I have checked that all tests pass, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#back-end-development).
- [ ] I have run all `pre-commit` hooks, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

**Comment by shamoon** [comment]
Thanks, we’ll take a look. One immediate thing: settings are already saved to the database, I think this should use that rather than a new Env var service.

**Comment by shamoon** [comment]
Changing this to draft, at least frontend code needs some work, so does the UI. Also (as you noted) the `pre-commit` hooks weren't run on the code so the style checks are failing. 

Definitely appreciate the work!

**Comment by shamoon** [comment]
Ok! Im happier with the frontend code 😅, please take a look @tim-vogel and let me know if you have any concerns. Screenshots of new styling are below for the curious

TODO (AFAIK):

- [ ] backend checks are failing
- [ ] frontend tests
- [ ] search?

Screen Shot 2022-08-07 at 9 37 13 PM Screen Shot 2022-08-07 at 9 36 59 PM

**Comment by tim-vogel** [comment]
@shamoon 

The new design looks great. What I miss on the screenshots is the possibility to delete a comment. 

Moving the setting to activate the comment function to the GUI is absolutely ok for me. The main thing is that the users who don't want this feature can disable it. 

With the open point of the search from the todo, I'm curious about the source code if someone implements this. I did not know how to implement this properly.

I will try to test the revised source code tonight.

Thanks for your revisions and corrections to make a clean source code out of my design. I'm sure it took you some time to correct everything.

**Comment by shamoon** [comment]
> The new design looks great. What I miss on the screenshots is the possibility to delete a comment.

The delete button shows only when you hover over a comment! =)

**Comment by tim-vogel** [comment]
I really like the feature with the fade-in delete button. The design is also very good. But what bothers me is that new comments are entered below the comment list.  For longer comment lists, you always have to scroll to the end of the comment list to enter a new comment. I think it would be better if the input field for new comments was placed at the top.

**Comment by shamoon** [comment]
> I think it would be better if the input field for new comments was placed at the top.

I debated this a bit, I think there are pros / cons to each choice. Like if you think users will be coming here more often to view comments or to add a new one, that sorta determines where it goes =) Visually I prefer at the bottom but open to discussion on this of course. I can imagine some visual tweaks that would help if others feel strongly it be moved

**Comment by mtettke123** [comment]
+1 for top and sorting comments newest (top) to oldest (bottom)

**Comment by tim-vogel** [comment]
Maybe it can be made adjustable. By means of CSS the order should be configurable. 

https://www.w3schools.com/cssref/css3_pr_order.asp

So the current view can remain as default. If the user wants to reverse the position, the order can be swapped with a CSS class.

Then another bool option in the settings and a CSS class would have to be created.

**Comment by shamoon** [comment]
Two separate things:

- Sorting of comments. Currently newest on top, dont think we need a toggle for this imho unless many feel so.
- "Add a comment" box on top vs bottom. Making this toggle-able just feels weird to me, like indecision. Let's get some input and try to reach a consensus. Ultimately there will probably always be opinions on both sides! =)

**Comment by tim-vogel** [comment]
Getting opinions is always a good idea ;-). Offering an option to leave the decision to the user I do not feel as indecisive. Rather, it offers flexibility and each user can set their own preferences. As you have already written, there will be lovers for each variant. 

But let the users decide how they want it ;-)

**Comment by tim-vogel** [comment]
I consider a setting for sorting the comments to be superfluous and not purposeful.

**Comment by enerschnutz** [comment]
> +1 for top and sorting comments newest (top) to oldest (bottom)

I second this

**Comment by torwag** [comment]
Talking about UX, would it be a good idea to have a symbol for comments in the matrix view and adding the newest comment in the list-view? So users do not have to go into the detail view to read the latest comment.

**Comment by Solverz-0** [comment]
Will comments be included in the Document exporter when exporting/backing up paperless-ngx?

**Comment by stumpylog** [comment]
> Will comments be included in the Document exporter when exporting/backing up paperless-ngx?

They will.

**Comment by tim-vogel** [comment]
> Talking about UX, would it be a good idea to have a symbol for comments in the matrix view and adding the newest comment in the list-view? So users do not have to go into the detail view to read the latest comment.

I think this is a very good idea. However, there will be some different views on this again. For this reason, I think that we should first discuss this extension in a discussion. In addition, we should get the comment function into a release first. Extensions can then be added later.

**Comment by NetHans** [comment]
The topic has been somewhat forgotten.
How do we want to continue at this point?

**Comment by shamoon** [comment]
Backend tests are still failing, needs backend review. (Also one open question about placement of add comment box)

**Comment by stumpylog** [comment]
The backend test failures need to be updated for a new item in the manifest.  And the export of comment and their contentneeds to be checked in tests.  Looking at the other code atm.

**Comment by torwag** [comment]
Back to the UI decision,  I feel that both add new box and sorting order are mutual connected.  You want the "add new"  box close to the newest comment,  so that adding a new comment it would be listed right there for you to see.  It would be odd to add a new comment only to scroll down (or up)  a long list of comments to see the one you just added.

**Comment by shamoon** [comment]
> Back to the UI decision, I feel that both, "add new" box and sorting order are mutual connected.

Yes this was exactly my reasoning. Clicking “add comment” and then it shows at the opposite end of the list seemed wrong. Since newest comment on top made more sense that’s where I put the form.

**Comment by shamoon** [comment]
Ok, Im good with moving the form to the top, made the change.

**Comment by stumpylog** [comment]
I've fixed the backend testing, added a basic verification that comments make it into the JSON, as well as fixing a minor issue with the format of the export.

**Review by stumpylog — APPROVED** [review]

**Comment by shamoon** [comment]
Uh oh I just realized this is targeted to the `main` branch but should be to `dev`. Sorry for not noting earlier. I will change it but likely needs a rebase

**Comment by stumpylog** [comment]
whoops, that's lost the backend testing and the export fix.

**Comment by shamoon** [comment]
> whoops, that's lost the backend testing and the export fix.

my bad, hope you have it locally. Im adding frontend tests too

**Comment by shamoon** [comment]
Ok added frontend tests and added back your lost commits @stumpylog , not sure why tests are still failing?

**Comment by stumpylog** [comment]
Ok, that should fix all the backend and such.

One thing I notice, when deleting a comment, I'm taken back to the overall documents screen, instead of the detail/comments tab where I deleted the comment from.  Is that the expected behavior?  My idea would be it simply removes the comment, maybe refreshing the page or data to show it gone?

**Comment by shamoon** [comment]
> One thing I notice, when deleting a comment, I'm taken back to the overall documents screen, instead of the detail/comments tab where I deleted the comment from.

Oh weird, thats a regression of some kind, let me look into it, thanks for all the backend test stuff!

**Review by shamoon — APPROVED** [review]
Phew! ok this should be good-to-go...

Maybe search can be added down the road =)

File: src/documents/views.py [hunk 1]
```diff
@@ -379,6 +380,63 @@
...
+        except Document.DoesNotExist:
+            raise Http404()
+
+        currentUser = request.user
+
+        if request.method == "GET":
+            try:
+                return Response(self.getComments(doc))
+            except Exception as e:
+                return Response({"error": str(e)})
```
[the remark this answers is not in the document]
Comment (stumpylog) [reply src/documents/views.py]: I'm inclined to dismiss these, though they'd be easy to silence by only logging the exception and returning the generic error dictionary.  Thoughts?

[the remark this answers is not in the document]
Comment (shamoon) [reply src/documents/views.py]: Yea totally defer to you. I don’t know how useful the actual errors would be in practice. I changed this to log the full error and show a more generic output to UI, mostly to satisfy this QL warning. Again up to you of course if you want to change more!

[the remark this answers is not in the document]
Comment (stumpylog) [reply src/documents/views.py]: Nope, hadn't noticed your changes when I made this, that's exactly what I had in mind.

**Comment by shamoon** [comment]
Thanks all, we can continue testing this feature in `dev` prior to 1.8.1

**Comment by christofkac** [comment]
I've tested this feature now and have seen, that the comments are not exported in the same directory as the documents but rather "only" in the manifest.json.
As the search option is not yet available I thought it would be good to export the comments also in e.g. a text or JSON file next to the document. With this, you could use a normal file manager to search for those.
What do you think?

**Pull request merged**